### DISCUSSION TITLE
preferences: ability to dynamically set the preference schema or default value

### DISCUSSION THREAD

**Discussion by vince-fugnitto** [eclipse-theia/theia#9141] [Q&A] [answered]
I'm wondering if there is a way to dynamically set the preference schema, for instance setting the `default` value of a preference based on the host operating system. Does anyone know if this is possible? I was thinking that we can set a preference for the `PreferenceScope.Default` but this seems to be a `no-op`.

https://github.com/eclipse-theia/theia/blob/f8bcc19eba884cbd64112aeb691e8c97aad87197/packages/core/src/browser/preferences/preference-contribution.ts#L329-L331

@kittaakos do you perhaps know of any way this can be done, perhaps you encountered the use-case previously.

**Comment by kittaakos** [comment]
No, I haven't done similar before but I am interested in how to make it work. What exactly is the requirement here?
Do you want to dynamically change the preferences schema or set and use a default value based on the OS? It was not clear to me.

If the latter, can't you do something like this:
```ts
import { isWindows } from '@theia/core/lib/common/os';
import { PreferenceSchema } from '@theia/core/lib/browser/preferences';
export const MySchema: PreferenceSchema = {
    type: 'object',
    properties: {
        'some.preference.key': {
            type: 'boolean',
            default: isWindows ? false : true
        }
    }
}
```

Or, you can always generate the final app's `package.json` and at build time you can give different preference values if you want to:
```jsonc
{
  // other things
  "theia": {
    "frontend": {
      "config": {
        "preferences": {
          "some.preference.key": false
        }
      }
    },
    // rest
}
```

I do not know if VS Code allows registering/removing configuration schemas dynamically, probably no. However, they allow using of preferences that are noop on certain OSs. For example, I can change the `terminal.integrated.env.linux` configuration on macOS, and it probably has no effect.

**Reply by vince-fugnitto** [reply]
> No, I haven't done similar before but I am interested in how to make it work. What exactly is the requirement here?

We have an intern who is working on supporting spawning external terminals for the application (electron version), and we are looking into setting a valid `default` value based on the user's host machine similarly to vscode. The issue is we need injected services and so on to define the preferences but this is not possible from what I understand.

Ideally we'd want to define a proper schema from the start, or a way to at least modify preference schemas (this API can perhaps be useful for the framework):
- https://github.com/microsoft/vscode/blob/9e4520310ed570357290e7c00e75ff0b483c4b1e/src/vs/workbench/contrib/externalTerminal/node/externalTerminal.contribution.ts#L65-L106

**Reply by kittaakos** [reply]
> his API can perhaps be useful for the framework

Sure, and for VS Code compatibility, these preferences are required in Theia.

**Reply by vince-fugnitto** [reply]
> > his API can perhaps be useful for the framework
> 
> Sure, and for VS Code compatibility, these preferences are required in Theia.

Correct, so perhaps this current limitation with having static schemas can be improved in the framework and hopefully cover more use-cases. Thanks for the insight!

**Comment by paul-marechal** [comment, marked as answer]
From what I understand, it is already supported to contribute preference schemas later. Without this VS Code extensions wouldn't be able to contribute preferences. Just call `PreferenceSchemaProvider.setSchema` at any point and it should be available from the UI to edit.

The issue is with our `PreferenceProxy` bindings... They currently don't support any asynchronous process, and they must be bound in advance, most likely before the schema is actually contributed. Binding such a proxy before contributing the schema shouldn't be an issue since they should only be "typed views" into the `PreferenceService` anyway.

I have a branch where I tried to make this change to the `PreferenceProxy` to support using a promise to a schema:
https://github.com/eclipse-theia/theia/compare/mp/lazy-preference-proxy

My approach was to make schema-dependent operations no-ops as long as the schema is not yet resolved. The assumption is that the preferences shouldn't be accessed before most components are initialized. Just like we would workaround the proxies by directly querying the `PreferenceService`.

Although while doing this change I found weird the amount of work the proxy does. It feels like it does too much and should rather delegate most of the logic to the `PreferenceService`?
